pkg/sources/generator/tickgen.go

```diff
@@ -217,42 +197,35 @@ func NewMemGen(
 		}
 	}
 
-	// attach a source publisher so the source can assign the watermarks.
-	genSrc.sourcePublishWM = genSrc.buildSourceWatermarkPublisher(publishWMStores)
+	// create a source watermark publisher
+	sourceWmPublisher := publish.NewSourcePublish(ctx, genSrc.pipelineName, genSrc.vertexName, publishWMStores,
+		publish.WithDelay(vertexInstance.Vertex.Spec.Watermark.GetMaxDelay()), publish.WithDefaultPartitionIdx(vertexInstance.Replica))
 
-	// we pass in the context to forwarder as well so that it can shut down when we cancelFn the context
-	forwarder, err := sourceforward.NewDataForward(vertexInstance, genSrc, writers, fsd, transformerApplier, fetchWM, genSrc, toVertexPublisherStores, idleManager, forwardOpts...)
+	// we pass in the context to sourceForwarder as well so that it can shut down when we cancelFn the context
+	sourceForwarder, err := sourceforward.NewDataForward(vertexInstance, genSrc, writers, fsd, transformerApplier, fetchWM, sourceWmPublisher, toVertexPublisherStores, idleManager, forwardOpts...)
 	if err != nil {
 		return nil, err
 	}
-	genSrc.forwarder = forwarder
+	genSrc.forwarder = sourceForwarder
 
 	return genSrc, nil
 }
 
-func (mg *memgen) buildSourceWatermarkPublisher(publishWMStores store.WatermarkStore) publish.Publisher {
-	// for tickgen, it can be the name of the replica
-	entityName := fmt.Sprintf("%s-%d", mg.vertexInstance.Vertex.Name, mg.vertexInstance.Replica)
-	processorEntity := entity.NewProcessorEntity(entityName)
-	// source publisher toVertexPartitionCount will be 1, because we publish watermarks within the source itself.
-	return publish.NewPublish(mg.lifecycleCtx, processorEntity, publishWMStores, 1, publish.IsSource(), publish.WithDelay(mg.vertexInstance.Vertex.Spec.Watermark.GetMaxDelay()))
-}
-
-func (mg *memgen) GetName() string {
+// GetName returns the name of the source
+func (mg *memGen) GetName() string {
 	return mg.vertexName
 }
 
-// GetPartitionIdx returns the partition number for the source vertex buffer
-// Source is like a buffer with only one partition. So, we always return 0
-func (mg *memgen) GetPartitionIdx() int32 {
-	return 0
+// Partitions returns the partitions for the source.
+func (mg *memGen) Partitions() []int32 {
+	return []int32{mg.vertexInstance.Replica}
```

Comment:
For single partition sources, we are using Replica as default index, would it be better if we just use `0`? It's better in a way that it explicitly shows `I am the first partition`.

Reply:
https://github.com/numaproj/numaflow/pull/1385#discussion_r1414901009